homeassistant/components/npi_gpio/binary_sensor.py

```diff
@@ -0,0 +1,89 @@
+"""Support for binary sensor using NPi GPIO."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components import npi_gpio
+from homeassistant.components.binary_sensor import (
+    BinarySensorDevice, PLATFORM_SCHEMA)
+from homeassistant.const import DEVICE_DEFAULT_NAME
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_BOUNCETIME = 'bouncetime'
+CONF_INVERT_LOGIC = 'invert_logic'
+CONF_PORTS = 'ports'
+CONF_PULL_MODE = 'pull_mode'
+
+DEFAULT_BOUNCETIME = 50
+DEFAULT_INVERT_LOGIC = False
+DEFAULT_PULL_MODE = 'UP'
+
+
+_SENSORS_SCHEMA = vol.Schema({
+    cv.positive_int: cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_PORTS): _SENSORS_SCHEMA,
+    vol.Optional(CONF_BOUNCETIME, default=DEFAULT_BOUNCETIME): cv.positive_int,
+    vol.Optional(CONF_INVERT_LOGIC, default=DEFAULT_INVERT_LOGIC): cv.boolean,
+    vol.Optional(CONF_PULL_MODE, default=DEFAULT_PULL_MODE): cv.string,
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Raspberry PI GPIO devices."""
+    pull_mode = config.get(CONF_PULL_MODE)
+    bouncetime = config.get(CONF_BOUNCETIME)
+    invert_logic = config.get(CONF_INVERT_LOGIC)
+
+    binary_sensors = []
+    ports = config.get('ports')
+    for port_num, port_name in ports.items():
+        binary_sensors.append(NPiGPIOBinarySensor(port_name, port_num,
+                                                  pull_mode, bouncetime,
+                                                  invert_logic))
+    add_entities(binary_sensors, True)
+
+
+class NPiGPIOBinarySensor(BinarySensorDevice):
+    """Represent a binary sensor that uses Raspberry Pi GPIO."""
+
+    def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
+        """Initialize the NPi binary sensor."""
+        self._name = name or DEVICE_DEFAULT_NAME
+        self._port = port
+        self._pull_mode = pull_mode
+        self._bouncetime = bouncetime
+        self._invert_logic = invert_logic
+        self._state = None
+
+        npi_gpio.setup_input(self._port, self._pull_mode)
```

Comment:
We want to avoid side effects in init method. Please move this call out of the class or into a dedicated setup method.

Reply:
I'm not able to fix that. I can't find a reference code or example to do that for that kind of component. Both Arduino and Beagle Bone GPIO's components (that i used as reference) use this setup.

Reply:
Add a method `setup` in the entity. Move this call to that method. Call the method on each entity after creating the entity before adding the entities.